sdks/python/apache_beam/runners/portability/fn_api_runner.py

```diff
@@ -1386,7 +1396,21 @@ def get_worker_handlers(self, environment_id, num_workers):
     # assume it's using grpc if environment is not EMBEDDED_PYTHON.
     if environment.urn != python_urns.EMBEDDED_PYTHON and \
         self._grpc_server is None:
-      self._grpc_server = GrpcServer(self._state, self._job_provision_info)
+      self._grpc_server = GrpcServer(
+          self._state, self._job_provision_info,
+          num_workers * len(self._environments))
+
+    # each gRPC server is running with fixed number of threads (num_workers
+    # * len(self._environments)), where num_workers is from the first call to
+    # get_worker_handlers(). Assumption here is a worker has a connection to a
+    # gRPC server. In case a stage tries to add more workers
+    # than the threads we have, some workers cannot attached to gRPC and
+    # pipeline will hang, so raise an error here.
+    if self._grpc_server is not None and \
+        num_workers > self._grpc_server.num_threads:
```

Comment:
This is not correct, the condition is `num_workers * len(self._environments) > ...`. Here is where naming would help. 

You can create a variable `max_total_workers = num_workers * len(self._environments)`. Then you can both pass this in and compare against it. 

(While we're at it, I would restructure the if statements here to be a bit clearer. E.g.

```
if environment.urn == python_urns.EMBEDDED_PYTHON:
  pass  # no need for a GrpcServer
elif self._grpc_server is None:
  self._grpc_server = ...
elif max_num_workers > self._grpc_server.max_workers:
  raise ...

Reply:
Thanks Robert, again, I learned a lot from you.